File: app/src/main/java/com/example/client/screens/profile/present/ProfilePresent.java
package com.example.client.screens.profile.present;

import com.example.client.api.ApiClient;
import com.example.client.api.service.UserService;
import com.example.client.app.MyFirebaseService;
import com.example.client.app.Preferences;
import com.example.client.models.message.MessageModel;
import com.example.client.models.profile.ProfileModel;
import com.example.client.screens.profile.fragment.IProfileView;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class ProfilePresent implements IProfilePresent {
    private IProfileView pView;
    private ProfileModel user;
    public ProfilePresent(IProfileView pView) {
        this.pView = pView;
        user = Preferences.getInstance().getProfile();
    }

    @Override
    public void onLogout() {
        MyFirebaseService myFirebaseService = new MyFirebaseService();
        myFirebaseService.deleteToken();
        UserService service = ApiClient.getInstance().create(UserService.class);
//        service.updateDeviceToken(user.getEmail(),null).enqueue(new Callback<MessageModel>() {
//            @Override
//            public void onResponse(Call<MessageModel> call, Response<MessageModel> response) {
//                if(response.body().isStatus()){
//                    Preferences.getInstance().deleteDeviceToken();
//                    Preferences.getInstance().deleteProfile();
//                    Preferences.getInstance().deleteAccessToken();
//                    Preferences.getInstance().deleteCart();
//                    pView.logout(response.body().isStatus());
//                }
//                else {
//                    pView.logout(false);
//                }
//            }
//
//            @Override
//            public void onFailure(Call<MessageModel> call, Throwable t) {
//                pView.logout(false);
//            }
//        });
    }

    @Override
    public void getUserFromRes() {
        pView.showUserInfo(Preferences.getInstance().getProfile());
    }
}

File: app/src/main/java/com/example/client/screens/profile/present/ProfilePresent.kt
package com.example.client.screens.profile.present

import com.example.client.app.Constants
import com.example.client.app.MyFirebaseService
import com.example.client.app.Preferences
import com.example.client.app.RxBus
import com.example.client.base.BasePresenterMVP
import com.example.client.screens.profile.fragment.IProfileView
import com.example.client.usecase.ProfileUseCase

class ProfilePresent(mView: IProfileView): BasePresenterMVP<IProfileView>(mView), IProfilePresent {
    private val profileUseCase by lazy { ProfileUseCase.newInstance() }
    private val preferences by lazy { Preferences.newInstance() }

    override fun bindData() {
        mView?.showUserInfo(preferences.profile)
    }

    override fun onLogout() {
        mView?.showLoading()
        val myFirebaseService = MyFirebaseService()
        myFirebaseService.deleteToken()
        subscribe(profileUseCase.updateDeviceToken(preferences.profile.email, ""),{
            mView?.run {
                hideLoading()
                if (it.is_error) {
                    showErrorMessage(getErrorMessage(it.code))
                    hideLoading()
                    return@subscribe
                }
                logout()
                preferences.deleteAll()
            }
        }, {
            it.printStackTrace()
            mView?.run {
                hideLoading()
                showErrorMessage(getErrorMessage(1001))
            }
        })
    }

    override fun navigateToContact() {
        mView?.toContactScreen(preferences.profile.id)
    }

    override fun onCompositedEventAdded() {
        super.onCompositedEventAdded()
        add(RxBus.newInstance().subscribe{
            when (it.key) {
                Constants.EventKey.UPDATE_PROFILE_AVATAR, Constants.EventKey.UPDATE_PROFILE_INFO -> bindData()
            }
        })
    }
}